### PR TITLE
Fix number parsing with negative sign

### LINKED ISSUE #47524: Swedish locale, NumberFormat.NegativeSign from ICU ignores parseLenient data from CLDR
### Description

Since .Net 5.0 in combination with Win10 May 2019 Update (build 18362) and later, ICU instead of NLS is used as source for locale data if I understand it correctly.

This causes issues with the negative sign using a Swedish locale, which then gets changed from Unicode 'Hyphen-Minus' to 'Minus Sign'. This actually does look good until you realize that you don't have 'Minus Sign' on a swedish keyboard, you only have 'Hyphen-Minus'. Users of our applications can therefore never enter negative numbers in any application since this change.

This would not be a problem if .Net 5.0 or the ICU library on windows respected the '[parseLenient](https://unicode-org.atlassian.net/browse/CLDR-10779?jql=text%20~%20%22sv%20minus%20sign%22)' data array that is associated with this from CLDR, as this includes the 'Hyphen-Minus'. However a strict equality seems to be used instead and 'Hyphen-Minus' can no longer be used to describe a negative number. 

Can this be fixed in the .Net runtime, or is this an underlying issue with the ICU library on Windows?

Related issue: #44678
Related issue: #40922

### Configuration
Win10 May 2019 Update (build 18362) and later.
with .Net 5.0 or later

### Workaround
Force usage of NLS instead of ICU
`<ItemGroup>`
`    <RuntimeHostConfigurationOption Include="System.Globalization.UseNls" Value="true" />`
`</ItemGroup>`

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes #47524

This change allow parsing negative numbers written with the hyphen character `-` when using cultures with negative sign not hyphen but it is one of the alternative Unicode minus sign characters. 
I have done some basic perf testing with this change and didn't notice any noticeable change.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Globalization/NumberFormatInfo.cs [hunk 1]
```diff
@@ -155,10 +161,23 @@ private static void VerifyDigitSubstitution(DigitShapes digitSub, string propert
         }
 
         internal bool HasInvariantNumberSigns => _hasInvariantNumberSigns;
+        internal bool AllowHyphenDuringParsing => _allowHyphenDuringParsing;
 
-        private void UpdateHasInvariantNumberSigns()
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (GrabYourPitchforks) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/NumberFormatInfo.cs:166]: Is this AI annotation necessary? This method should only be called on construction or when the property is set, both of which should be very rare.

File: src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs [hunk 2]
```diff
@@ -1725,6 +1757,8 @@ internal static unsafe ParsingStatus TryParseDecimal(ReadOnlySpan<char> value, N
             return ParsingStatus.OK;
         }
 
+        internal static bool SpanStartsWith(ReadOnlySpan<char> span, char c) => span.Length > 0 && span[0] == c;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs:1760]: FWIW, the JIT today will generate better code if you instead write this as:
```suggestion
        internal static bool SpanStartsWith(ReadOnlySpan<char> span, char c) => !span.IsEmpty && span[0] == c;
```
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGBLAOwwwo/bABsm5JA2AQIEgMoAHbPwUZsUDLgDqvDAAtyACgBKMbABMA8vzEBPZaoA8YA5oB8DXCv5oGbpoBAJQMALxePqoAdAAyMPwA5oYMXgAMDABkmd6+ANppALrhYQEA3Fw8AkIi4pLSsvIMTmoaWrr6BqRmFjZ2jr6u7lCRvv6BUCHhXgCEUfzRAJK4AKL4Shj2WTnzBcVhpWAV1AC+QA===

File: src/libraries/System.Private.CoreLib/src/System/Globalization/NumberFormatInfo.cs [hunk 3]
```diff
@@ -155,10 +161,23 @@ private static void VerifyDigitSubstitution(DigitShapes digitSub, string propert
...
+        private void InitializeInvariantAndNegativeSignFlags()
         {
             _hasInvariantNumberSigns = _positiveSign == "+" && _negativeSign == "-";
+            _allowHyphenDuringParsing = _negativeSign.Length == 1 &&
+                                        _negativeSign[0] switch {
```
Comment (GrabYourPitchforks) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/NumberFormatInfo.cs:171]: Nit: add code comment explaining where this test came from. If we wanted to, we could even have a unit test that cycles through this entire list and ensures a hyphen matches every one of them. We have some unit test helper libraries that parse the Unicode data files, used primarily for Rune and StringInfo and S.T.E.W, but if there's interest we could also adapt it for this.

(Whether we should have a unit test that iterates through this really depends on whether this list is populated from one of the Unicode data files.)

**Review by GrabYourPitchforks — COMMENTED** [review]
Do we also need to update __runtime/src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs__? There are some references to `NumberFormatInfo.NegativeSign` in that file.

Other files to inspect:
* __System/Windows/Forms/NumericUpDown.cs__
* __System/Text/Json/Serialization/Converters/Value/EnumConverter.cs__

(That latter file stores current culture data into a static field, which is probably a bug.)

File: src/libraries/System.Private.CoreLib/src/System/Globalization/NumberFormatInfo.cs [hunk 4]
```diff
@@ -71,6 +73,10 @@ public sealed class NumberFormatInfo : IFormatProvider, ICloneable
 
         private bool _hasInvariantNumberSigns = true;
 
+        // When _allowHyphenDuringParsing is set to true, will allow the number parser to accept the hyphen `-` U+002D as a negative sign even if the culture
+        // negative sign is not the hyphen. Good example of that is Swedish culture (e.g. "sv-SE") which has U+2212 as negative sign.
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/NumberFormatInfo.cs:77]:
```suggestion
        // When _allowHyphenDuringParsing is set to true, we'll allow the number parser to accept the hyphen `-` U+002D as a negative sign even if the culture
        // negative sign is not the hyphen. For example, the Swedish culture (e.g. "sv-SE") has U+2212 as the negative sign.
```

File: src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs [hunk 5]
```diff
@@ -295,7 +295,7 @@ private static unsafe bool TryParseNumber(ref char* str, char* strEnd, NumberSty
                 // "-Kr 1231.47" is legal but "- 1231.47" is not.
                 if (!IsWhite(ch) || (styles & NumberStyles.AllowLeadingWhite) == 0 || ((state & StateSign) != 0 && ((state & StateCurrency) == 0 && info.NumberNegativePattern != 2)))
                 {
-                    if ((((styles & NumberStyles.AllowLeadingSign) != 0) && (state & StateSign) == 0) && ((next = MatchChars(p, strEnd, info.PositiveSign)) != null || ((next = MatchChars(p, strEnd, info.NegativeSign)) != null && (number.IsNegative = true))))
+                    if ((((styles & NumberStyles.AllowLeadingSign) != 0) && (state & StateSign) == 0) && ((next = MatchChars(p, strEnd, info.PositiveSign)) != null || ((next = MatchNegativeSignChars(p, strEnd, info)) != null && (number.IsNegative = true))))
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs:298]: BigInteger has an old copy of the parsing code:
https://github.com/dotnet/runtime/blob/master/src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs
There's an issue to merge them again (https://github.com/dotnet/runtime/issues/28657), but until that happens, presumably we need to update that as well?

File: src/libraries/System.Private.CoreLib/src/System/Globalization/NumberFormatInfo.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/NumberFormatInfo.cs:166]: I'll remove the annotation.

**Comment by tarekgh** [comment]
@GrabYourPitchforks 

> Do we also need to update runtime/src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs? 

Yes, you are right. I'll fix this one too. I am not sure why we have parsing code in multiple places. 

> System/Windows/Forms/NumericUpDown.cs

This file is in WinForms repo and not runtime. I looked at it and they have `keyInput.Equals(negativeSign)` if they need this work correctly they can convert it to ` keyInput.Equals(negativeSign, StringComparison.CurrentCulture)`.

> System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

This should be fine because they are testing with the linguistic StartsWith which should work.

File: src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs (hunk 5 above)
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs:298]: I am seeing this file is used in System.Runtime.Numeric only. it is even weird that this stored in the common folder. I guess that was the file we used in the globalization parsing before and it didn't get chance to be cleaned up. anyway, I'll add the fix to this file too.

**Comment by tarekgh** [comment]
Any other comments here?

**Review by GrabYourPitchforks — APPROVED** [review]
LGTM. Though I'd still like to see a comment on the switch statement saying "here's where we pulled this list of values from."

**Comment by GrabYourPitchforks** [comment]
@tarekgh I sent an IM to you on Teams asking for clarification on something. Check that if you have a sec before merging.

**Comment by danmosemsft** [comment]
Do we need to create a followup issue track reducing duplication you discovered here?

**Comment by tarekgh** [comment]
I just discovered there some changes I didn't submit yet :-( so I am going to submit it soon.

**Comment by GrabYourPitchforks** [comment]
@danmosemsft There's a tracking issue at https://github.com/dotnet/runtime/issues/28657.

File: src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs [hunk 6]
```diff
@@ -328,6 +329,35 @@ private static bool IsWhite(char ch)
                 return null;
             }
 
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static unsafe bool AllowHyphenDuringParsing(NumberFormatInfo info)
+            {
+                string negativeSign = info.NegativeSign;
+                return negativeSign.Length == 1 &&
+                       negativeSign[0] switch {
+                           '\u2012' or         // Figure Dash
```
Comment (danmosemsft) [inline src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs:338]: I think I'm missing something, shouldn't this return true for regular hyphen \u002d? Even if it's not functionally necessary, it seems like a confusing method name if it doesn't return true for that.

Comment (danmosemsft) [reply src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs:338]: Or maybe AllowAlternativeHyphenDuringParsing might be a better name?

Comment (tarekgh) [reply src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs:338]: No. If we have the negative sign as '-' then we don't need to do anything. The fix here is to allow parsing when using hyphen in the parsed string. when we have NegativeSign as hyphen it is already working today without any fix.

Comment (tarekgh) [reply src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs:338]: > Or maybe AllowAlternativeHyphenDuringParsing might be a better name?

No, we are allowing parsing the hyphen and not the alternative hyphens. 

The fix here is when having a culture using alternative hyphens as negative sign, we'll allow parsing strings using the hyphen 0+002D `-`.

Comment (danmosemsft) [reply src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs:338]: Ah, you're quite right. I was thinking backwards.

**Comment by tarekgh** [comment]
Now all changes is submitted.

**Comment by GrabYourPitchforks** [comment]
Thanks for adding the comments and tests! 👍

**Pull request merged**